benchmark/UnsortedJoinBenchmark.cpp

```diff
@@ -0,0 +1,96 @@
+// Copyright 2015, University of Freiburg,
+// Chair of Algorithms and Data Structures.
+// Author: Andre Schlegel (November of 2022, schlegea@informatik.uni-freiburg.de)
+// Author of the file this file is based on: Björn Buchhold (buchhold@informatik.uni-freiburg.de)
+
+#include <gtest/gtest.h>
+
+#include <cstdio>
+#include <algorithm>
+
+#include "engine/Engine.h"
+#include "engine/Join.h"
+#include "engine/QueryExecutionTree.h"
+#include "util/Forward.h"
+#include "engine/idTable/IdTable.h"
+#include "../benchmark/Benchmark.h"
+
+// Local helper function.
+#include "../benchmark/util/IdTableHelperFunction.h"
+
+// Non-local helper function.
+#include "../test/util/IdTableHelpers.h"
+#include "../test/util/JoinHelpers.h"
+#include "../test/IndexTestHelpers.h"
+
+// Benchmarks for unsorted tables, with and without overlapping values in
+// IdTables. Done with normal join and hash join.
+void BM_UnsortedIdTable(BenchmarkRecords* records) {
+  // For easier changing of the IdTables size.
+  // Important: Apparently Join::join has an inbuild timeout checker, which
+  // keeps getting triggerd by tables with a size between 500-600, or higher.
+  // This leads to a segmentation fault, because it keeps trying to throw an
+  // exception about the time out, but that exception tries to include
+  // information about the query execution tree using a shared pointer. And, as
+  // far as I can tell, that pointer was never actually initialized here, so
+  // it points into memory, which we are not allowed to access. Which leads
+  // to the seg fault.
+  const size_t NUMBER_ROWS = 500;
+  const size_t NUMBER_COLUMNS = NUMBER_ROWS; 
+
+  Join J{Join::InvalidOnlyForTestingJoinTag{}, ad_utility::testing::getQec()};
+  DISABLE_WARNINGS_CLANG_13
+  auto hashJoinLambda = [&J]<int A, int B, int C>(auto&&... args) {
+    ENABLE_WARNINGS_CLANG_13
+    return J.hashJoin(AD_FWD(args)...);
+  };
+  DISABLE_WARNINGS_CLANG_13
+  auto joinLambda = [&J]<int A, int B, int C>(auto&&... args) {
+    ENABLE_WARNINGS_CLANG_13
+    return J.join<A, B, C>(AD_FWD(args)...);
+  };
```

Comment:
This code is now duplicated three times.

Maybe this can be in your `JoinHelpers`, a useful patterns is writing a function that returns a lambda:

```
auto makeHashJoinLambda() {
  Join J{Join::InvalidOnlyForTestingJoinTag{}, ad_utility::testing::getQec()};
    return  [J = std::move(J)]<int A, int B, int C>(auto&&... args) {
      ENABLE_WARNINGS_CLANG_13
      return J.hashJoin(AD_FWD(args)...);
    };
)
}

Reply:
Same for the `joinLambda` (two times in the benchmark, and one time in the test).

Reply:
I implemented it in [3f307e0](https://github.com/ad-freiburg/qlever/pull/860/commits/3f307e075c90229954e01dfe1cceb3520a31c1a2), but with one change: The creation and initialization of J happens inside the lambda.

I changed it like that, because, after leaving `makeHashJoinLambda` J life time is over and I'm not sure, if the lambda will always work without it. It seems to, in my small tests, but I'm not 100% sure, that behavior is deterministic.

Reply:
Implemented everything correctly in [e353b69](https://github.com/ad-freiburg/qlever/pull/860/commits/e353b6902cdd4ef6342deca8ef8f300db19a0e96).